lib/chef/resource/homebrew_install.rb

```diff
@@ -0,0 +1,148 @@
+#
+# Copyright:: Copyright (c) Chef Software Inc.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+require_relative "../resource"
+
+class Chef
+  class Resource
+    class HomebrewInstall < Chef::Resource
+      unified_mode true
+
+      provides :homebrew_install
+
+      description "Use the **homebrew_install** resource to install the Homebrew package manager on macOS systems."
+      introduced "16.8"
+      examples <<~DOC
+      **Install Homebrew using the Internet to download Command Line Tools for Xcode**:
+      ```ruby
+      homebrew_install 'Install Homebrew and xcode command line tools if necessary' do
+        user 'someuser'
+        action :install
+      end
+      ```
+      **Install Homebrew using a customer-managed source to download Command Line Tools for Xcode from**:
+      ```ruby
+      homebrew_install 'Install Homebrew and xcode command line tools if necessary' do
+        xcode_tools_url 'https://somewhere.something.com/downloads/command_line_tools.dmg'
+        xcode_tools_pkg_name 'Command Line Tools.pkg'
+        user 'someuser'
+        action :install
+      end
+      ```
+      DOC
+
+      property :xcode_tools_url, String,
+        description: "A url pointing to a 'Command Line Tools for Xcode' dmg file"
+
+      property :xcode_tools_pkg_name, String,
+        description: "The name of the pkg inside the dmg located at the xcode_tools_url"
+
+      property :brew_source_url, String,
+        description: "A url pointing to a Homebrew installer",
+        default: "https://codeload.github.com/Homebrew/brew/zip/master"
+
+      property :user, String,
+        description: "The user to install Homebrew as. Note: Homebrew cannot be installed as root.",
+        required: true
+
+      action :install do
+        # Avoid all the work in the below resources if homebrew is already installed
+        return if ::File.exist?("/usr/local/bin/brew")
+
+        # check if 'user' is root and raise an exception if so
+        if Etc.getpwnam(new_resource.user).uid == 0
+          msg = "You are attempting to install Homebrew as Root. This is not permitted by Homebrew. Please run this as a standard user with admin rights"
+          raise Chef::Exceptions::InsufficientPermissions, msg
+        end
+
+        USER_HOME = Dir.home(new_resource.user).freeze
```

Comment:
we should make these lowercase since they're not really constants and that gets confusing

Reply:
done